CONTRIBUTING.md

```diff
@@ -3,16 +3,91 @@
 This project is [Apache 2.0 licensed](LICENSE) and accepts contributions via
 GitHub pull requests.
 
-## Guidelines
+## Contributing Guidelines
 
-* A change should not being merged unless it passes CI or there is a comment/update saying what testing was passed.
-* PRs should not be merged unless positively reviewed.
-* If a change is adding a feature, it should require a change to the README.md and the review should catch this.
-* If a change is a fix, it should have an issue. The review should make sure the comments state the issue (not just the number) and it should use the keywords that will close the issue on merge.
+* Unit tests are required to accompany all logic being introduced into the `pkg/` directory.
+* Functional tests are required to accompany any logic that modifies `pkg/apis/` API structs.
+* Follow the tide review flow for merging commits. This flow forces a review and all required CI jobs to pass before a PR is merged.
+* Read and understand the design principles outlined in this document that pertain to the code you are contributing to.
 
 ## Certificate of Origin
 
 By contributing to this project you agree to the Developer Certificate of
 Origin (DCO). This document was created by the Linux Kernel community and is a
 simple statement that you, as a contributor, have the legal right to make the
 contribution. See the [DCO](DCO) file for details.
+
+# Reconcile Loop Design Principles
+
+* Never block reconcile execution. ever. This means __no sleeps and no retry loops__
+* All __logic must be idempotent__. This means discovering what has already occurred by inspecting informer cache, and only mutating to converge on desired state.
+* __Never perform an apiserver List()__ request within the reconcile loop. This action is very apiserver intensive. Instead use informers for retrieving items from cache.
+* __Avoid mutating a cluster object the loop didn't create__. A exception to this would to be adding annotations to an existing object. Mutating a object's Spec that our controller didn't create should be avoided.
+* __Avoid performing an apiserver Get()__ request within the reconcile loop if at all possible. Again use informers. There are rare exceptions.
+* Use __finalizers__ in order to perform any cleanup tasks required before a CR is completely removed from etcd.
+* Use __owner references__ on any objects created and owned by a CR. Owner references ensure objects are automatically garbage collected after a corresponding CR is deleted.
```

Comment:
following the recent issues in CNV: maybe add that namespaced resource can have namespaced or clusterwide resources as owner, but clusterwide resources are only allowed to have clusterwide owners? Too easy to forget, and doing it wrong results in "magically" at random time deleted resources.